feat: Add opaque type support to PrestoSerializer

kunigami opened the pull request:
Summary:
The idea of opaque types in Presto serializer is that we'll serialize the underlying opaque type to StringView. This requires a serde callbacks to be registered via

  OpaqueType::registerSerialization();

then inside `PrestoSerializer.cpp` we use these to write and read values of type opaque.

Differential Revision: D64362525

pedroerp commented on velox/serializers/tests/PrestoSerializerTest.cpp:
```diff
@@ -1511,6 +1512,39 @@ INSTANTIATE_TEST_SUITE_P(
+
+  static std::shared_ptr<Foo> create(int64_t id) {
+    // Return the same instance if the id is already in the map, to make
+    // the operator== work with the instance before and after serde.
```
not sure I followed. Why do you need to compare that they are the same instances, and not just different instances which are equal?

kunigami replied:
I think it's because the comparison of shared_ptr<Foo> will compare the raw pointer address, so it needs to be the same instance. we could change the test utilities to de-reference before comparisons if desired.

bikramSingh91 commented on velox/serializers/PrestoSerializer.cpp:
```diff
@@ -1935,6 +2001,39 @@ void serializeFlatVector<TypeKind::BOOLEAN>(
+        continue;
+      }
+      stream->appendNonNull();
+      std::shared_ptr<void> ptr = rawValues[offset];
```
nit: unused variable

bikramSingh91 commented on velox/serializers/PrestoSerializer.cpp:
```diff
@@ -1935,6 +2001,39 @@ void serializeFlatVector<TypeKind::BOOLEAN>(
 }
 
+template <>
+void serializeFlatVector<TypeKind::OPAQUE>(
```
This implementation is used by the `PrestoBatchVectorSerializer` only. You also need to add an implementation for `PrestoIterativeVectorSerializer`. I am not sure which one you use for your use-case. To give you more context the `PrestoBatchVectorSerializer` is optimized for serializing full vectors whereas `PrestoIterativeVectorSerializer` is used for serializing subsets of rows scattered within a single vector(which is usually used for a partitioning exchange).

the `serializeFlatVector` implementation for `PrestoIterativeVectorSerializer` is here: https://github.com/facebookincubator/velox/blob/a33e8d733d9ed8f23121ec2737a55c56a9f740a8/velox/serializers/PrestoSerializer.cpp#L2737

Finally, for a complete implementation, you also need to implement `estimateSerializedSize` for both types of serializers. This is used by the PartitionedOutput node to figure out when it needs to flush data to the wire.

bikramSingh91 commented on velox/serializers/PrestoSerializer.cpp:
```diff
@@ -625,6 +628,65 @@ void read<StringView>(
+  }
+
+  const int64_t rawStringSizeBytes = dataSize * sizeof(char);
+  char* rawString = (char*)pool->allocate(rawStringSizeBytes);
```
nit: instead of directly using the pool's allocate API that returns a raw pointer, I would suggest create a buffer and using a BufferPtr instead since its much more safer as the space will be freed when BufferPtr is destroyed, otherwise if any of the steps throw after this then it can introduce a memory leak.

Its like using a shared_ptr vs a raw ptr and has similar benefits.

You can use:
```
BufferPtr newBuffer = AlignedBuffer::allocate<char>(dataSize, pool());
```

bikramSingh91 commented on velox/serializers/tests/PrestoSerializerTest.cpp:
```diff
@@ -1555,6 +1589,7 @@ TEST_F(PrestoSerializerTest, deserializeSingleColumn) {
       DOUBLE(),
       VARCHAR(),
       TIMESTAMP(),
+      OPAQUE<Foo>(),
```
This is just adding a type to `deserializeSingleColumn` which is not used for serialization across the wire (See method comment of `deserializeSingleColumn` for more info). Therefore, you would also want to add test coverage for the serialize (both batch and iterative) and deserialize methods.

Also would have to add tests for `estimateSerializedSize` when you implement it

kunigami commented:
addressing comments
- use `AlignedBuffer::allocate()`
- discussed w/ [bikramSingh91](https://github.com/bikramSingh91) that we don't need `PrestoIterativeVectorSerializer` and `estimateSerializedSize`
- add test case for opaque type serialization/deserialization

bikramSingh91 commented on velox/serializers/PrestoSerializer.cpp:
```diff
@@ -625,6 +628,64 @@ void read<StringView>(
+    const SerdeOpts&,
+    VectorPtr& result) {
+  // Opaque values are serialized by first converting them to string
+  // them serializing them as if they were string. The deserializable
```
nit: "then" serializing

bikramSingh91 commented on velox/serializers/tests/PrestoSerializerTest.cpp:
```diff
@@ -1339,6 +1373,25 @@ TEST_P(PrestoSerializerTest, encodedRoundtrip) {
+  auto inputVector = makeFlatVector<std::shared_ptr<void>>(
+      2,
+      [](vector_size_t row) { return Foo::create(row + 10); },
+      /*isNullAt=*/nullptr,
```
can you also add a test with nulls for completeness

bikramSingh91 commented on velox/serializers/tests/PrestoSerializerTest.cpp:
```diff
@@ -1339,6 +1373,25 @@ TEST_P(PrestoSerializerTest, encodedRoundtrip) {
   }
 }
 
+TEST_P(PrestoSerializerTest, opaqueBatchVectorSerializer) {
```
can you also add a test to ensure that opaque type will throw an error if used with Iterative serializer and if estimateSerializedSize is used

bikramSingh91 reviewed (commented):
Looks good. Just a few nits. The build is failing, can you please take a look into it

kunigami commented:
* address comments. use find() instead of contains() so it doesn't rely on newer c++ features
